tests/StatsWriterTest.php

```diff
@@ -0,0 +1,214 @@
+
+        $this->assertDatabaseHas('stats_events', [
+            'value' => 1,
+            'type' => 'change',
+            'created_at' => now()->subWeek(),
+        ]);
+    }
+
+    /** @test */
+    public function it_can_create_events_for_increments()
+    {
+        StatsWriter::for(StatsEvent::class)->increase();
+
+        $this->assertDatabaseHas('stats_events', [
+            'value' => 1,
+            'type' => 'change',
+        ]);
+    }
+
+    /** @test */
+    public function it_can_create_events_for_decrements()
+    {
+        StatsWriter::for(StatsEvent::class)->decrease();
+
+        $this->assertDatabaseHas('stats_events', [
+            'value' => -1,
+            'type' => 'change',
+        ]);
+    }
+
+    /** @test */
+    public function it_can_create_events_for_decrements_at_a_given_timestamp()
+    {
+        StatsWriter::for(StatsEvent::class)->decrease(1, now()->subWeek());
+
+        $this->assertDatabaseHas('stats_events', [
+            'value' => -1,
+            'type' => 'change',
+            'created_at' => now()->subWeek(),
+        ]);
+    }
+
+    /** @test */
+    public function it_can_create_events_for_setting_fixed_values()
+    {
+        StatsWriter::for(StatsEvent::class)->set(1337);
+
+        $this->assertDatabaseHas('stats_events', [
+            'value' => 1337,
+            'type' => 'set',
+        ]);
+    }
+
+    /** @test */
+    public function it_can_create_events_for_setting_fixed_values_at_a_given_timestamp()
+    {
+        StatsWriter::for(StatsEvent::class)->set(1337, now()->subWeek());
+
+        $this->assertDatabaseHas('stats_events', [
+            'value' => 1337,
+            'type' => 'set',
+            'created_at' => now()->subWeek(),
+        ]);
+    }
+
+//    /** @test */
+//    public function it_can_get_a_stats_query_object()
+//    {
+//        $query = OrderStats::query();
+//
+//        $this->assertInstanceOf(StatsQuery::class, $query);
+//        $this->assertInstanceOf(OrderStats::class, $query->getStatistic());
+//    }
+
+//    /** @test */
+//    public function it_can_get_the_value_at_a_given_time()
+//    {
+//        OrderStats::set(3, now()->subDays(19));
+//        OrderStats::decrease(1, now()->subDays(4));
+//        OrderStats::increase(3, now()->subDays(2));
+//
+//        $this->assertEquals(0, StatsQuery::for(OrderStats::class)->getValue(now()->subDays(30)));
+//        $this->assertEquals(3, StatsQuery::for(OrderStats::class)->getValue(now()->subDays(18)));
+//        $this->assertEquals(5, StatsQuery::for(OrderStats::class)->getValue(now()));
+//    }
+//
+//    /** @test */
+//    public function it_can_generate_and_array_of_periods()
+//    {
+//        $periods = StatsQuery::for(OrderStats::class)->start(now()->subYear())->end(now())->generatePeriods();
+//
+//        $this->assertCount(53, $periods);
+//
+//        $this->assertEquals([
+//            Carbon::parse('2018-12-31'),
```

Comment:
Are these old tests that still need fixing or can they be removed?

Reply:
Good spot! Saw that yesterday after re-reviewing my own stuff and almost forgot about it. 

I will update those. Perhaps I will move visibility to protected for `generatePeriods` as it's only used internally. I will add a set of new tests to cover this via `StatsWriter::get()`

Reply:
Awesome, thanks!